app/services/user-agent-service/datomstore.js

```diff
@@ -0,0 +1,239 @@
+        ':db/unique': ':db.unique/identity',
+        ':db/doc': 'A page\'s URL.',
+        ':db.install/_attribute': ':db.part/db',
+      },
+
+      'page/title': {
+        ':db/ident': ':page/title',
+        ':db/cardinality': ':db.cardinality/one',      // We supersede as we see new titles.
+        ':db/doc': 'A page\'s title.',
+        ':db.install/_attribute': ':db.part/db',
+      },
+
+      'event/visit': {
+        ':db/valueType': ':db.type/ref',
+        ':db/cardinality': 'db.cardinality/many',
+        ':db/doc': 'A visit to the page.',
+      },
+
+      // TODO: model sessions.
+      'visit/instant': {
+        ':db/cardinality': 'db.cardinality/one',
+        ':db/doc': 'When the visit occurred.',
+        ':db/index': true,
+        // ':db/valueType': 'db.type/instant',        // Not supported by Datascript.
+      },
+
+      'session/startedFromAncestor': {
+        ':db/valueType': 'db.type/ref',     // To a session.
+        ':db/cardinality': 'db.cardinality/one',
+        ':db/doc': 'The ancestor of a session.',
+      },
+
+      'session/startedInScope': {
+        ':db/valueType': 'db.type/string',
+        ':db/cardinality': 'db.cardinality/one',
+        ':db/doc': 'The parent scope of a session.',
+      },
+
+      // Numeric for now.
+      'session/startReason': {
+        ':db/cardinality': 'db.cardinality/many',
+        ':db/doc': 'The start reasons of a session.',
+      },
+
+      // Numeric for now.
+      'session/endReason': {
+        ':db/cardinality': 'db.cardinality/many',
+        ':db/doc': 'The end reasons of a session.',
+      },
+
+      'session/startTime': {
+        ':db/cardinality': 'db.cardinality/one',
+      },
+      'session/endTime': {
+        ':db/cardinality': 'db.cardinality/one',
+      },
+    });
+  }
+
+  static async open(dir) {
+    return new ProfileDatomStorage(dir);
+  }
+
+  close() {
+    return true;
+  }
+
+  async startSession(scope, ancestor, now = microtime.now(), reason = SessionStartReason.newTab) {
+    const result = datascript.core.transact_BANG_(this.conn, vector(
+      vector(DB_ADD, -1, 'session/startedInScope', `${scope}`),
+      vector(DB_ADD, -1, 'session/startedFromAncestor', ancestor),
+      vector(DB_ADD, -1, 'session/startReason', reason),
+      vector(DB_ADD, -1, 'session/startTime', now)
+    ));
+
+    return datascript.core.resolve_tempid(result.tempids, -1);
+  }
+
+  async endSession(session, now = microtime.now(), reason = SessionEndReason.tabClosed) {
+    return datascript.core.transact_BANG_(this.conn, vector(
+      vector(DB_ADD, session, 'session/endReason', reason),
+      vector(DB_ADD, session, 'session/endTime', now)
+    ));
+  }
+
+  // Chains through place ID.
+  starPage(url, session, action, now = microtime.now()) {
+    const result = datascript.core.transact_BANG_(this.conn, vector(
+      vector(DB_ADD, -1, 'page/url', url),
+    ));
+
+    // TODO
+
+    return datascript.core.resolve_tempid(result.tempids, -1);
```

Comment:
I'm not sure this `result.tempids` does what you think it does.  If you used `datascript.js`, maybe -- but I think this is a mori `map`, not a JS object.